source/s3_client.c

```diff
@@ -2054,3 +2054,159 @@ struct aws_byte_cursor aws_s3_meta_request_resume_token_upload_id(
+            /* We only interested in request succeed */
+            ++stats->num_upload_requests_succeed;
+            if (stats->num_upload_requests_succeed <= 10) {
+                /* Gether the data */
+                uint64_t request_time_ns =
+                    metrics->time_metrics.receive_end_timestamp_ns - metrics->time_metrics.send_start_timestamp_ns;
+                stats->request_time_sum_ns += request_time_ns;
+                ++stats->number_request_time;
+                if (stats->num_upload_requests_succeed == 10) {
+                    /* Decide we need a timeout or not */
+                    uint64_t average_request_time_ns = stats->request_time_sum_ns / stats->number_request_time;
+                    if (average_request_time_ns >= s_upload_timeout_threshold_ns) {
+                        /* We don't need a timeout, as retry will be slower than just wait for the server to response */
+                        stats->stop_timeout = true;
+                    } else {
+                        /* Start the timeout by 1 secs */
+                        aws_atomic_store_int(&client->upload_timeout_ms, 1000);
+                    }
+                }
+                goto unlock;
+            }
+            /* Starts to update timeout on case of succeed */
+            ++stats->num_rate_track_completed;
+            /* Response to first byte is time taken for the first byte data received from the request finished
+             * sending */
+            uint64_t response_to_first_byte_time_ns =
+                metrics->time_metrics.receive_start_timestamp_ns - metrics->time_metrics.send_end_timestamp_ns;
+            stats->response_to_first_byte_time_ns_sum += response_to_first_byte_time_ns;
+            ++stats->response_to_first_byte_time_ns_number;
+
+            uint64_t average_response_to_first_byte_time_ns =
+                stats->response_to_first_byte_time_ns_sum / stats->response_to_first_byte_time_ns_number;
+            uint64_t expected_timeout_ns =
+                average_response_to_first_byte_time_ns +
+                aws_timestamp_convert(g_expect_timeout_offset_ms, AWS_TIMESTAMP_MILLIS, AWS_TIMESTAMP_NANOS, NULL);
+            double timeout_ns_double = (double)current_timeout_ns * 0.99 + (double)expected_timeout_ns * 0.01;
+            updated_timeout_ns = (uint64_t)timeout_ns_double;
+            break;
+        case AWS_ERROR_HTTP_RESPONSE_FIRST_BYTE_TIMEOUT:
+            if (stats->num_upload_requests_succeed < 10) {
+                goto unlock;
+            }
+
+            /* Starts to update timeout on case of timed out */
+            ++stats->num_rate_track_completed;
+            ++stats->num_rate_track_timed_out;
+
+            size_t timeout_threshold = stats->num_rate_track_completed / 100;
+            if (stats->num_rate_track_completed % 100 > 0) {
+                /* Use the ceiling of 1% of made request as threshold */
+                ++timeout_threshold;
```

Comment:
this is weird, if completed is 1000 then we do NOT ++, but if completed is 999 or 1001 we DO ++

maybe use floating point math, instead of stressing about integer division

Reply:
failure_rate = (double)num_failed / num_completed;
if failure_rate > 0.01:
    asdf
elif failure_rate > 0.001:
    qwer

Reply:
used `ceil` instead to be less confusing